tests/test_copy_async.py

```diff
@@ -972,6 +972,105 @@ async def test_copy_table_across(aconn_cls, dsn, faker, mode):
+    execution_log = []
+
+    async def copy_task():
+        """Copy task that writes two rows with controlled pauses."""
+        async with cur.copy("copy copy_concurrency_test from stdin") as copy:
+            # Pause after entering copy context
+            execution_log.append("entered_copy")
+            copy_entered.set()
+            await can_proceed.wait()
+
+            # Write first row and pause
+            await copy.write_row((1, "first"))
+            execution_log.append("wrote_row_1")
+            wrote_first.set()
+            await can_proceed.wait()
+
+            # Write second row and pause
+            await copy.write_row((2, "second"))
+            execution_log.append("wrote_row_2")
+            wrote_second.set()
+            await can_proceed.wait()
+
+        # Copy context exited, lock should now be released
+        execution_log.append("exited_copy")
+
+    async def worker_task(worker_id):
+        """
+        Worker that attempts to execute a query on a different cursor.
+        Should block until copy completes due to connection lock.
+        """
+        # Try to execute on another cursor - this should block until copy exits
+        worker_cur = aconn.cursor()
+        await worker_cur.execute("select 1")
+        execution_log.append(f"worker_{worker_id}_completed")
+
+    # Start the copy task
+    t_copy = spawn(copy_task, ())
+
+    # Wait for copy to enter, then spawn first worker
+    await copy_entered.wait()
+    t_worker1 = spawn(worker_task, (1,))
+
+    # Allow copy to proceed to write first row
+    can_proceed.set()
+    can_proceed.clear()
+    await wrote_first.wait()
+
+    # Spawn second worker after first row
+    t_worker2 = spawn(worker_task, (2,))
+
+    # Allow copy to proceed to write second row
+    can_proceed.set()
+    can_proceed.clear()
+    await wrote_second.wait()
+
+    # Spawn third worker after second row
+    t_worker3 = spawn(worker_task, (3,))
+
+    # Allow copy to exit
+    can_proceed.set()
+
+    # Wait for all tasks to complete
+    await gather(t_copy, t_worker1, t_worker2, t_worker3)
+
+    # Verify the data was written correctly
+    await cur.execute("select * from copy_concurrency_test order by id")
+    rows = await cur.fetchall()
+    assert rows == [(1, "first"), (2, "second")]
+
+    # Verify that all workers completed AFTER copy exited
+    # This confirms the connection lock was held throughout the copy operation
+    copy_exit_idx = execution_log.index("exited_copy")
+    worker1_idx = execution_log.index("worker_1_completed")
+    worker2_idx = execution_log.index("worker_2_completed")
+    worker3_idx = execution_log.index("worker_3_completed")
+
+    assert copy_exit_idx < worker1_idx, "Worker 1 should complete after copy exits"
+    assert copy_exit_idx < worker2_idx, "Worker 2 should complete after copy exits"
+    assert copy_exit_idx < worker3_idx, "Worker 3 should complete after copy exits"
```

Comment:
Pytest gives a meaningful error message when an assert fails. Maybe this would be enough to check?

```suggestion
    want = "exited_copy worker_1_completed worker_2_completed worker_3_completed".split()
    assert execution_log[:-4] == want
```

Reply:
That's perfectly good, thank you. It is what I would have actually written if I had written the test myself :)